README.md
Jira Cookbook
==================
[![Chef cookbook](https://img.shields.io/cookbook/v/jira.svg)](https://supermarket.chef.io/cookbooks/jira)
[![Build Status](https://secure.travis-ci.org/afklm/jira.png?branch=master)](http://travis-ci.org/afklm/jira)

*WARNING* - This cookbook was essentially replaced by a new cookbook starting version 2.0. This is a breaking change.

## Description

Installs/Configures Atlassian JIRA. Please see [COMPATIBILITY.md](COMPATIBILITY.md) for more information about JIRA releases that are tested and supported by this cookbook and its versions.

## Requirements

### Chef

* Chef 11+ for version 2.0.0+ of this cookbook

### Platforms

* CentOS 6,7
* RHEL 6
* Ubuntu 12.04

### Databases

* Microsoft SQL Server
* MySQL
* Postgres

### Cookbooks

Required [Opscode Cookbooks](https://github.com/opscode-cookbooks/)

* [apache2](https://github.com/opscode-cookbooks/apache2) (if using apache2 recipe)
* [ark](https://github.com/opscode-cookbooks/ark)
* [database](https://github.com/opscode-cookbooks/database) (if using database recipe)
* [mysql](https://github.com/opscode-cookbooks/mysql) (if using database recipe with MySQL)
* [postgresql](https://github.com/opscode-cookbooks/postgresql) (if using database recipe with Postgres)

Required Third-Party Cookbooks

* [mysql_connector](https://github.com/bflad/chef-mysql_connector) (if using MySQL database)

Suggested [Opscode Cookbooks](https://github.com/opscode-cookbooks/)

* [java](https://github.com/opscode-cookbooks/java)
* [tomcat](https://github.com/opscode-cookbooks/tomcat)

### JDK/JRE

The Atlassian JIRA Linux installer will automatically configure a bundled JRE. If you wish to use your own JDK/JRE, with say the `java` cookbook, then as of this writing it must be Oracle and version 1.6 ([Supported Platforms](https://confluence.atlassian.com/display/JIRA/Supported+Platforms))

Necessary configuration with `java` cookbook:
* `node['java']['install_flavor'] = "oracle"`
* `node['java']['oracle']['accept_oracle_download_terms'] = true`
* `recipe[java]`

A /ht to [@seekely](https://github.com/seekely) for the [documentation nudge](https://github.com/bflad/chef-jira/issues/2).

## Attributes

These attributes are under the `node['jira']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
checksum | SHA256 checksum for JIRA install | String | auto-detected by helper method
home_path | home directory for JIRA | String | /var/atlassian/application-data/jira
install_path | location to install JIRA | String | /opt/atlassian/jira
install_type | JIRA install type - "installer" or "standalone" | String | installer
init_type | JIRA init service type - "sysv" | String | sysv
url | URL for JIRA install | String | auto-detected by helper method
user | user running JIRA | String | jira
version | JIRA version to install | String | 6.1.5

**Notice:** If `['jira']['install_type']` is set to `installer`, then the installer will try to upgrade your JIRA instance located in `['jira']['install_path']` (if it exists) to the `['jira']['version']`.

If you want to avoid an unexpected upgrade, just set or override `['jira']['version']` attribute value to that of your current JIRA version.

### JIRA Database Attributes

All of these `node['jira']['database']` attributes are overridden by `jira/jira` encrypted data bag (Hosted Chef) or data bag (Chef Solo), if it exists

Attribute | Description | Type | Default
----------|-------------|------|--------
host | FQDN or "localhost" (localhost automatically installs `['database']['type']` server in default recipe) | String | localhost
name | JIRA database name | String | jira
password | JIRA database user password | String | changeit
type | JIRA database type - "mssql", "mysql", or "postgresql" | String | mysql
user | JIRA database user | String | jira

### JIRA JVM Attributes

These attributes are under the `node['jira']['jvm']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
minimum_memory | JVM minimum memory | String | 512m
maximum_memory | JVM maximum memory | String | 768m
maximum_permgen | JVM maximum PermGen memory | String | 256m
java_opts | additional JAVA_OPTS to be passed to JIRA JVM during startup | String | ""
support_args | additional JAVA_OPTS recommended by Atlassian support for JIRA JVM during startup | String | ""

### JIRA Tomcat Attributes

These attributes are under the `node['jira']['tomcat']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
port | Tomcat HTTP port | Fixnum | 8080

## Recipes

* `recipe['jira']` 'Installs/configures Atlassian JIRA'
* `recipe['jira::apache2']` 'Installs/configures Apache 2 as proxy (ports 80/443)'
* `recipe['jira::container_server_configuration']` 'Configures container server for JIRA deployment'
* `recipe['jira::database']` 'Installs/configures MySQL/Postgres server, database, and user for JIRA'
* `recipe['jira::installer']` 'Installs/configures JIRA via installer'
* `recipe['jira::standalone']` 'Installs/configures JIRA via standalone archive'

## Usage

### JIRA Server Data Bag

Optionally for securely overriding attributes on Hosted Chef, create a `jira/jira` encrypted data bag with the model below. Chef Solo can override the same attributes with a `jira/jira` unencrypted data bag of the same information.

_required:_
* `['database']['type']` - "mssql", "mysql", or "postgresql"
* `['database']['host']` - FQDN or "localhost" (localhost automatically installs `['database']['type']` server)
* `['database']['name']` - Name of JIRA database
* `['database']['user']` - JIRA database username
* `['database']['password']` - JIRA database username password

_optional:_
* `['database']['port']` - Database port, defaults to standard database port for `['database']['type']`

Repeat for other Chef environments as necessary. Example:

    {
      "id": "jira",
      "development": {
        "database": {
          "type": "postgresql",
          "host": "localhost",
          "name": "jira",
          "user": "jira",
          "password": "jira_db_password",
        }
      }
    }

### Default JIRA Installation

The simplest method is via the default recipe, which uses `node['jira']['install_type']` (defaults to installer).

* Optionally (un)encrypted data bag or set attributes
  * `knife data bag create jira`
  * `knife data bag edit jira jira --secret-file=path/to/secret`
* Add `recipe[jira]` to your node's run list.

### Standalone JIRA Installation

Operates similarly to installer installation, however has added benefits of using `ark` to create version symlinks of each install. Easily can rollback upgrades by changing `node['jira']['version']`.

* Optionally (un)encrypted data bag or set attributes
  * `knife data bag create jira`
  * `knife data bag edit jira jira --secret-file=path/to/secret`
* Set `node['jira']['install_type']` to standalone
* Add `recipe[jira]` to your node's run list.

### JIRA WAR Support

Starting from JIRA 7, WAR installation is no longer supported:
https://confluence.atlassian.com/jira/installing-jira-war-185729447.html

### Custom JIRA Configurations

Using individual recipes, you can use this cookbook to configure JIRA to fit your environment.

* Optionally (un)encrypted data bag or set attributes
  * `knife data bag create jira`
  * `knife data bag edit jira jira --secret-file=path/to/secret`
* Add individual recipes to your node's run list.

## Testing and Development

* Quickly testing with Vagrant: [VAGRANT.md](VAGRANT.md)
* Full development and testing workflow with Test Kitchen and friends: [TESTING.md](TESTING.md)

For Vagrant, you may need to add the following hosts entries:

* 192.168.50.10 jira-centos-6
* 192.168.50.10 jira-ubuntu-1204
* (etc.)

The running JIRA server is then accessible from the host machine:

CentOS 6 Box:
* Web UI (installer/standalone): https://jira-centos-6/

Ubuntu 12.04 Box:
* Web UI (installer/standalone): https://jira-ubuntu-1204/

## Contributing

Please see contributing information in: [CONTRIBUTING.md](CONTRIBUTING.md)

## Maintainers

* KLM Royal Dutch Airlines

## License

Please see licensing information in: [LICENSE](LICENSE)


CHANGELOG.md
## x.y.z (pending)

* Set collation to 'C' for PostgreSQL database
* Added cloud testing config for `test-kitchen` via DigitalOcean.
  [[GH-55]](https://github.com/afklm/jira/issues/55)
* Force apache restart to circumvent `mod_proxy` bug on Ubuntu 12.04.
  [[GH-52]](https://github.com/afklm/jira/issues/52)
* Use versioned ark install paths.
  [[GH-56]](https://github.com/afklm/jira/issues/56)
* Small fixups to make all test suites succeed on cloud platforms.
  [[GH-57]](https://github.com/afklm/jira/issues/57)
* Fix standalone directory perms and accompanying CentOS bug.
  [[GH-57]](https://github.com/afklm/jira/issues/57)
* Lock `postgresql` cookbook to v3.4.16 so that its derived attributes
  don't mess with our setting of psql version.
  [[GH-60]](https://github.com/afklm/jira/issues/60)
* Added tests for proper psql version.
  [[GH-60]](https://github.com/afklm/jira/issues/60)

## 2.7.3

* Added JIRA 7.0.2
* Bumped default version to 7.0.2

## 2.7.2

* Added sensitive true for crowd_sso template
* Set Postgres DB owner to JIRA user
* Explicitly set `home_path` perms.
  [[#48]](https://github.com/afklm/jira/issues/48)

## 2.7.0

* Added support for JIRA 7.0
* Added `node['jira']['flavor']` attribute with default value of 'software'
* Bump default JIRA version to 7.0.0

## 2.6.3

* Add JIRA 6.4.12 and bump default version

## 2.6.2

* Added a `node['jira']['group']` attribute for clarity and override ability
* Issue warn in library for unsupported DB types to make override possible

## 2.6.1

* Removed unused config entries from cookbook metadata
* Replaced file cookbook with FileEdit#search_file_replace

## 2.6.0

* Redirect http-based requests to https
* Auto upgrade if `node['jira']['version']` is higher than installed version
* Configures Crowd SSO if `node['jira']['crowd_sso']['enabled']` is true
* Improve file restrictions for standalone installs

## 2.5.1

* MIGRATED: renamed cookbook chef_jira -> jira after getting the supermarket
            namespace
* Bump default version to JIRA 6.4.11
* Use https for Jira downloads.
  [[GH-18]](https://github.com/afklm/jira/issues/18)
* Added service restart when ark resource changes.
  [[GH-16]](https://github.com/afklm/jira/issues/16)
* Added support for PostgreSQL 9.2+ in `dbconfig.xml`.
  [[GH-14]](https://github.com/afklm/jira/issues/14)
* Fixed LWRPs after cookbook name change.
  [[GH-13]](https://github.com/afklm/jira/pull/13)
* Set Tomcat `proxyName`/`proxyPort` even without SSL.
  [[GH-11]](https://github.com/afklm/jira/issues/11)
* Removed unnecessary non-dynamic `web.xml` template.
  [[GH-10]](https://github.com/afklm/jira/issues/10)
* Fixed setting of `jira.home` for all install types.
  [[GH-15]](https://github.com/afklm/jira/issues/15)

Thanks go to @elijah @gsreynolds @legal90 and @patcon for helping out in this
release.

## 2.1.0

* MIGRATED: renamed cookbook jira -> chef_jira

## 2.0.1

* Bugfix: #8: Remove include_attribute from default attributes to prevent hard dependency on tomcat cookbook

## 2.0.0

If you were using the default recipe, there are no changes you need to make for your environment to upgrade this cookbook.

Major features are full support for standalone deployments and war building/deployments. Using ark where possible (its worked well for those using my Stash cookbook).

I've removed the upgrade recipe in favor of using ark. If you'd like to keep old JIRA installations around (in case of upgrade issues, etc.), I would recommend switching to the standalone install_type since ark will automatically create versioned symlinks and you can easily revert `node['jira']['version']`. To convert install_type if install_path is the default (X.Y.Z being currently installed `node['jira']['version']`):
* `service jira stop`
* `mv /opt/atlassian/jira /opt/atlassian/jira-X.Y.Z`
* `ln -s /opt/atlassian/jira-X.Y.Z /opt/atlassian/jira`
* Set `node['jira']['install_type']` to standalone
* Run Chef Client

Other than that, migrated some recipes/templates and split out some recipes from the old linux_installer recipe, so ensure that if you're using a custom run list or template override for any nodes, that they include the new recipes/template location as necessary.

Full details:
* REMOVED: upgrade recipe and associated backup_* and *_backup attributes
* MIGRATED: linux_installer -> installer recipe
* MIGRATED: tomcat_configuration -> container_server_configuration recipe
* MIGRATED: Tomcat templates into tomcat folder:
  * permgen.sh.erb -> tomcat/permgen.sh.erb
  * server.xml.erb -> tomcat/server.xml.erb
  * setenv.sh.erb -> tomcat/setenv.sh.erb
  * web.xml.erb -> tomcat/web.xml.erb
* SPLIT: SysV init service configuration into sysv recipe and add init_type attribute
* SPLIT: database jar deployment (mysql_connector_j, etc.) into container_server_jars recipe which also installs JIRA jars for war install_type
* Bugfix: Use :create action instead of :create_if_missing for installer remote_file
* Enhancement: Add standalone and war recipes and quite a few attributes for supporting those install_type's
* Enhancement: Add build_war recipe
* Enhancement: LWRPs for handling multiple instances of install, etc.
* Enhancement: Bump default JIRA version to 6.1.5

## 1.7.0

* Bump default JIRA version to 6.1

## 1.6.0

* Bump default JIRA version to 6.0.7

## 1.5.0

* Bump default JIRA version to 6.0.6

## 1.4.0

* Initial Microsoft SQL Server support

## 1.3.0

* Bump default JIRA version to 6.0.5

## 1.2.0

* Bump default JIRA version to 6.0.2

## 1.1.0

* Bump default JIRA version to 6.0.1

## 1.0.0

* Split default recipe into individual recipes
* apache2 recipe does not include default recipe
* Load database/tomcat settings via Jira.settings library (bonus: help support Chef Solo)
* Moved apache2 attributes into default attributes
* Bump default JIRA version to 5.2.11
* Added url_base attribute
* Auto-detect checksum attribute for some versions
* Added Vagrantfile and Test Kitchen for testing
* minitest fixes

attributes/default.rb
default['jira']['home_path']          = '/var/atlassian/application-data/jira'
default['jira']['install_path']       = '/opt/atlassian/jira'
default['jira']['install_type']       = 'installer'
default['jira']['version']            = '7.0.0'
default['jira']['flavor']             = 'software'
default['jira']['user']               = 'jira'
default['jira']['group']              = 'jira'
default['jira']['backup_when_update'] = false
default['jira']['ssl']                = false

# Defaults are automatically selected from version via helper functions
default['jira']['url']      = nil
default['jira']['checksum'] = nil

default['jira']['apache2']['access_log']         = ''
default['jira']['apache2']['error_log']          = ''
default['jira']['apache2']['port']               = 80

# Defaults are automatically selected from fqdn and hostname via helper functions
default['jira']['apache2']['virtual_host_name']  = nil # node['fqdn']
default['jira']['apache2']['virtual_host_alias'] = nil # node['hostname']


default['jira']['apache2']['ssl']['access_log']       = ''
default['jira']['apache2']['ssl']['error_log']        = ''
default['jira']['apache2']['ssl']['chain_file']       = ''
default['jira']['apache2']['ssl']['port']             = 443

case node['platform_family']
when 'rhel'
  default['jira']['apache2']['ssl']['certificate_file'] = '/etc/pki/tls/certs/localhost.crt'
  default['jira']['apache2']['ssl']['key_file']         = '/etc/pki/tls/private/localhost.key'
else
  default['jira']['apache2']['ssl']['certificate_file'] = '/etc/ssl/certs/ssl-cert-snakeoil.pem'
  default['jira']['apache2']['ssl']['key_file']         = '/etc/ssl/private/ssl-cert-snakeoil.key'
end

default['jira']['database']['host']     = '127.0.0.1'
default['jira']['database']['name']     = 'jira'
default['jira']['database']['password'] = 'changeit'
default['jira']['database']['type']     = 'mysql'
default['jira']['database']['user']     = 'jira'

# Needed for postgresql unfortunately
if node['jira']['database']['type'] == 'postgresql'
  case node['platform_family']
  when 'debian'
    default['apt']['compile_time_update'] = true
  end
end

# Default is automatically selected from database type via helper function
default['jira']['database']['port'] = nil

default['jira']['jvm']['minimum_memory']  = '256m'
default['jira']['jvm']['maximum_memory']  = '768m'
default['jira']['jvm']['maximum_permgen'] = '256m'
default['jira']['jvm']['java_opts']       = ''
default['jira']['jvm']['support_args']    = ''

default['jira']['tomcat']['port'] = '8080'

default['jira']['crowd_sso']['enabled']        = false
default['jira']['crowd_sso']['sso_appname']    = 'jira'
default['jira']['crowd_sso']['sso_password']   = 'changethistosomethingsensible'
default['jira']['crowd_sso']['crowd_base_url'] = 'http://localhost:8095/crowd/'
